extensions/tn_cache/tn_cache.go

```diff
@@ -68,6 +129,162 @@ func init() {
+		return fmt.Errorf("app.DB is not a sql.DB")
+	}
+
+	tx, err := db.BeginTx(ctx.TxContext.Ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback(ctx.TxContext.Ctx)
+
+	// Check if stream is configured and has been refreshed
+	var configuredFromTime *int64
+	var lastRefreshed *string
+
+	result, err := tx.Execute(ctx.TxContext.Ctx, `
+		SELECT from_timestamp, last_refreshed
+		FROM ext_tn_cache.cached_streams
+		WHERE data_provider = $1 AND stream_id = $2
+	`, dataProvider, streamID)
+	if err != nil {
+		return fmt.Errorf("failed to query cached_streams: %w", err)
+	}
+
+	if len(result.Rows) == 0 {
+		// Stream not configured for caching
+		return resultFn([]any{false})
+	}
+
+	row := result.Rows[0]
+	if row[0] != nil {
+		t := row[0].(int64)
+		configuredFromTime = &t
+	}
+	if row[1] != nil {
+		s := row[1].(string)
+		lastRefreshed = &s
+	}
+
+	// If stream hasn't been refreshed yet, no cached data
+	if lastRefreshed == nil {
+		return resultFn([]any{false})
+	}
+
+	// Check if requested from_time is within our cached range
+	if configuredFromTime != nil && fromTime < *configuredFromTime {
+		// Requested time is before what we have cached
+		return resultFn([]any{false})
+	}
+
+	// At this point, we know the stream is configured and has been refreshed
+	// Check if we actually have events in the requested range
+	var eventCount int64
+	if toTime == nil {
+		// No upper bound specified
+		result, err = tx.Execute(ctx.TxContext.Ctx, `
+			SELECT COUNT(*) FROM ext_tn_cache.cached_events
+			WHERE data_provider = $1 AND stream_id = $2 AND event_time >= $3
+		`, dataProvider, streamID, fromTime)
+	} else {
+		// Upper bound specified
+		result, err = tx.Execute(ctx.TxContext.Ctx, `
+			SELECT COUNT(*) FROM ext_tn_cache.cached_events
+			WHERE data_provider = $1 AND stream_id = $2 AND event_time >= $3 AND event_time <= $4
+		`, dataProvider, streamID, fromTime, *toTime)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to count cached events: %w", err)
+	}
+
+	if len(result.Rows) > 0 {
+		eventCount = result.Rows[0][0].(int64)
+	}
+
+	return resultFn([]any{eventCount > 0})
+}
+
+// handleGetCachedData handles the get_cached_data precompile method
+func handleGetCachedData(ctx *common.EngineContext, app *common.App, inputs []any, resultFn func([]any) error) error {
+	// Extract parameters
+	dataProvider := strings.ToLower(inputs[0].(string))
+	streamID := inputs[1].(string)
+	fromTime := inputs[2].(int64)
+
+	var toTime *int64
+	if len(inputs) > 3 && inputs[3] != nil {
+		t := inputs[3].(int64)
+		toTime = &t
+	}
+
+	db, ok := app.DB.(sql.DB)
+	if !ok {
+		return fmt.Errorf("app.DB is not a sql.DB")
+	}
+
+	tx, err := db.BeginTx(ctx.TxContext.Ctx)
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer tx.Rollback(ctx.TxContext.Ctx)
+
+	var result *sql.ResultSet
+	if toTime == nil {
+		// No upper bound specified
+		result, err = tx.Execute(ctx.TxContext.Ctx, `
```

Comment:
The only thing missing here to be totally equal to ours is to have the anchor value if the first returned value isn't exact on the lower bound

not sure if you prefer directly on the SQL to be more efficient, or at a separate call as it's fast anyway (just a small roundtrip)